### PR TITLE
WIP refactor(controller): job-step-task refactor

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description
demo: http://jobrefactor.pre.intra.starwhale.ai/
refactor : 
1.[**code structure**]Tune code logic: job(scheduler\hotJobHolder\service)->step->task(scheduler\log\watcher)
> move jobUpdateHelper to jobService
> rename jobLoader to jobOperator(add schedule\cancel\pause\remove function)
> move statuses which in stepHelper to StepStatusCalculator

2.[**code structure**]Encapsulate mappers into their respective domains and expose services
3.Tune StateMachine impl
> remove unused status
> add some case for statemachine

4.Fix some bug: 
> add judgement at JobEventHandler
> cancel running task when job is fail
> fix that the canceled task status can be updated to PREPARING 

TODOs(future to do):
- [ ] model serving refactor
- [ ] k8sScheduler refactor

## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java:1]: Do we have a business case where we schedule jobs? Now we only schedule tasks.

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                        WatchableTaskFactory watchableTaskFactory,
+                        TaskScheduler taskScheduler,
+                        TaskService taskService) {
+        this.jobHolder = jobHolder;
+        this.watchableTaskFactory = watchableTaskFactory;
+        this.taskScheduler = taskScheduler;
+        this.taskService = taskService;
+    }
+
+    public Job schedule(@NotNull Job job, Boolean resumePausedOrFailTasks) {
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java:69]: I think `load` is better than the word `schedule` and keep it in `jobloader` class

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java [hunk 2]
```diff
@@ -27,13 +27,17 @@
  */
 public interface HotJobHolder {
 
-    void adopt(Job job);
+    void add(Job job);
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java:30]: what is the benefit of renaming this method

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java [hunk 3]
```diff
@@ -27,13 +27,17 @@
...
 
-    Collection<Job> ofIds(Collection<Long> ids);
+    Collection<Job> getJobs(Set<Long> ids);
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java:32]: what is the benefit of renaming this method

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/split/JobSpliteratorImpl.java [hunk 4]
```diff
@@ -16,20 +16,19 @@
...
 package ai.starwhale.mlops.domain.job.split;
 
-import ai.starwhale.mlops.common.util.BatchOperateHelper;
 import ai.starwhale.mlops.domain.job.JobDao;
 import ai.starwhale.mlops.domain.job.bo.Job;
 import ai.starwhale.mlops.domain.job.spec.JobSpecParser;
 import ai.starwhale.mlops.domain.job.spec.StepSpec;
 import ai.starwhale.mlops.domain.job.status.JobStatus;
-import ai.starwhale.mlops.domain.job.step.mapper.StepMapper;
+import ai.starwhale.mlops.domain.job.step.StepService;
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/split/JobSpliteratorImpl.java:24]: what is the benefit of forcing JobSpliterator to depend on service level, do we have a proper abstraction of service layer. 
What is the benefit of increasing hierarchy in our code withou proper abstractions.

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/status/StepStatusMachine.java [hunk 5]
```diff
@@ -24,32 +24,33 @@
...
-    public boolean couldTransfer(StepStatus statusNow, StepStatus statusNew) {
+    public static final Set<StepStatus> FINAL_STATUS = Set.of(FAIL, SUCCESS, CANCELED);
+
+    public static boolean couldTransfer(StepStatus statusNow, StepStatus statusNew) {
         return transferMap.get(statusNow).contains(statusNew);
     }
 
-    public boolean isFinal(StepStatus status) {
-        return transferMap.get(status).isEmpty();
+    public static boolean isFinal(StepStatus status) {
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/status/StepStatusMachine.java:52]: How to mock a StepStatusMachine when we do UT if you change it to static

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/status/JobStatusCalculator.java [hunk 6]
```diff
@@ -16,66 +16,61 @@
...
+        map.put(JobStatus.SUCCESS, Set.of(
+            new StatusRequirement<>(Set.of(StepStatus.SUCCESS), RequireType.ALL),
+            new StatusRequirement<>(Set.of(StepStatus.SUCCESS), RequireType.MUST)
+        ));
+
+        stepStatusRequirementSetMap = Collections.unmodifiableMap(map);
     }
 
-    public JobStatus desiredJobStatus(Collection<StepStatus> stepStatuses) {
+    public static JobStatus desiredJobStatus(Collection<StepStatus> stepStatuses) {
```
Comment (anda-ren) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/job/status/JobStatusCalculator.java:73]: How to mock a JobStatusCalculator when we do UT if you change it to static

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    public void insert(StepEntity entity) {
+        stepMapper.insert(entity);
+    }
+
+    public void updateLastStep(Long id, Long lastId) {
+        stepMapper.updateLastStep(id, lastId);
+    }
+
+    public void fillJobSteps(Job job) {
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java:59]: there is the reason why move fillJobSteps to StepService:

> [code structure]Encapsulate mappers into their respective domains and expose services

and this describtion is not correct,(it not effect the jobBo and job would have full properties, see:https://github.com/goldenxinxing/starwhale/blob/0abd5554610794ae149b67a1eeeff02cf933dd40/server/controller/src/main/java/ai/starwhale/mlops/domain/job/converter/JobBoConverter.java#L153):

> If this method is moved from JobBoConverter, you assume that there is somewhere that job has no steps in the runtime and people should call this fillJobSteps method when job has no steps.

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        steps.forEach(step -> {
+            Long nextStepId = linkMap.get(step.getId());
+            if (null == nextStepId) {
+                return;
+            }
+            step.setNextStep(stepMap.get(nextStepId));
+        });
+    }
+
+    public void updateStepStatus(Step step, StepStatus newStatus) {
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java:102]: > [code structure]Encapsulate mappers into their respective domains and expose services
so taskWatcher is better not holder a stepMapper but a stepService

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/init/HotJobsLoader.java
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/init/HotJobsLoader.java:1]: `init` is just a init functions package which belong to job domain, so put the cache init func can also say the past

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/JobLoader.java
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/JobLoader.java:1]: emm. JobLoader has not only one duty, it wrap tasks, put job to HotJobHolder and schedule task to k8s, and the duty is not express the meaning of load.
the second question: Do we have a real business logic which schedules a job? yes, actually, we schedule job first, then schedule task in k8s and use event to drive the status change( task-> step->job)，and job service have the need of cancel\resume\remove\pause， so call it JobScheduler is fit

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java (hunk 2 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java:30]: It's just a bit hard to see the effect at a glance

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/split/JobSpliteratorImpl.java (hunk 4 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/split/JobSpliteratorImpl.java:24]: other domain's mapper should not directly used by other domain

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/JobLoader.java
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/JobLoader.java:1]: renamed to JobOperator

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/status/JobUpdateHelper.java
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/status/JobUpdateHelper.java:1]: removed it to JobOperator

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java]: recoverd to the origin impl, you are right, jobBoConverter need the only one to export the full entity

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            step.setFinishTime(now.getTime());
+            stepMapper.updateFinishedTime(step.getId(), now);
+        }
+        if (StepStatus.RUNNING == newStatus) {
+            step.setStartTime(now.getTime());
+            stepMapper.updateStartedTime(step.getId(), now);
+        }
+    }
+
+    public Step firsStep(List<Step> linkedSteps) {
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/StepService.java]: done

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java:1]: fixed

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java (hunk 1 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java]: renamed to addAndSchedule

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/status/StepStatusMachine.java (hunk 5 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/step/status/StepStatusMachine.java:52]: see: https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#48

File: server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java (hunk 3 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/job/cache/HotJobHolder.java]: The form ofxxx has been uniformly used

**Pull request closed without merging**